lib/spack/spack/util/imp/importlib_importer.py

```diff
@@ -45,4 +44,9 @@ def load_source(full_name, path, prepend=None):
     """
     # use our custom loader
     loader = PrependFileLoader(full_name, path, prepend)
-    return loader.load_module()
+    # Note this is a ModuleSpec object defined in importlib and not a
+    # Spack Spec object
+    module_spec = spec_from_loader(full_name, loader)
+    mod = module_from_spec(module_spec)
+    module_spec.loader.exec_module(mod)
+    return mod
```

Comment:
> Would you say these lines (49-52) are the key change that addresses the problem?

Yes

I don't precisely know why the issue comes up, but https://github.com/spack/spack/issues/20005#issuecomment-731496593 and https://github.com/spack/spack/issues/20005#issuecomment-731707332 explain the most about the problem (and why it might be intermittent in your case). Those results suggested a conflict between our manual importing mechanism and `import spack.hooks.sbang`, so I examined our manual import mechanism for any issues (at which point I noticed we were using a deprecated function as part of doing manual imports).

Does your usage have to do with binary caching (in that case it might be an instance of what is described in https://github.com/spack/spack/issues/20005#issuecomment-731707332)? If not, then are you able to reproduce it by retrying several times in a row?

Reply:
Thanks for this background information and pointers to the most relevant comments.  My most recent encounter with it did have to do with binary caching (only indirectly in that the module was imported), and was fixed by cherry-picking the commit mentioned in [#20005 (comment)](https://github.com/spack/spack/issues/20005#issuecomment-824093790).  But if I encounter it again I'll keep in my to try several times in a row.

Reply:
@scheibelp I'm reading this right now. It seems we are using a way to import modules that is advised for Python 3.4 or earlier. Also, I was not aware of duplication here, but there's a similar function I coded for #8612:

https://github.com/spack/spack/blob/2473e0cc82e9be4a4e04177343792c22b23a0d30/lib/spack/llnl/util/lang.py#L804-L833

that can probably be reused to get rid of most of the boilerplate in the `spack.util.imp` module.

Reply:
There's a caveat I think we currently miss, but I need to double check:

> The module will exist in sys.modules before the loader executes the module code. This is crucial because the module code may (directly or indirectly) import itself; adding it to sys.modules beforehand prevents unbounded recursion in the worst case and multiple loading in the best

The excerpt is taken from [here](https://docs.python.org/3/reference/import.html#loading) and I think we don't add our module to `sys.module` beforehand.